static/app/components/core/layout/flex.tsx

```diff
@@ -1,38 +1,325 @@
+
+  ${p => rc('width', p.width, p.theme)};
+  ${p => rc('min-width', p.minWidth, p.theme)};
+  ${p => rc('max-width', p.maxWidth, p.theme)};
+
+  ${p => rc('height', p.height, p.theme)};
+  ${p => rc('min-height', p.minHeight, p.theme)};
+  ${p => rc('max-height', p.maxHeight, p.theme)};
 
-interface FlexProps {
-  align?: CSSProperties['alignItems'];
-  direction?: CSSProperties['flexDirection'];
-  flex?: CSSProperties['flex'];
-  gap?: Space | ReturnType<typeof space>;
   /**
-   * Determines whether the flex container should be displayed as an inline-flex.
+   * This cast is required because styled-components does not preserve the generic signature of the wrapped component.
+   * By default, the generic type parameter <T> is lost, so we use 'as unknown as' to restore the correct typing.
+   * https://github.com/styled-components/styled-components/issues/1803
    */
-  inline?: boolean;
-  justify?: CSSProperties['justifyContent'];
-  wrap?: CSSProperties['flexWrap'];
+` as unknown as <T extends ContainerElement = 'div'>(
+  props: ContainerProps<T>
+) => React.ReactElement;
+
+// It is unfortunate, but Emotion seems to use the fn callback name in the classname, so lets keep it short.
+function rc<T>(
+  property: string,
+  value: Responsive<T> | undefined,
+  theme: Theme,
+  // Optional resolver function to transform the value before it is applied to the CSS property.
+  resolver?: (value: T, theme: Theme) => string | number
+): SerializedStyles | undefined {
+  if (!value) {
+    return undefined;
+  }
+
+  // Most values are unlikely to be responsive, so we can resolve
+  // them directly and return early.
+  if (!isResponsive(value)) {
+    // prettier-ignore
+    return css`${property}: ${resolver ? resolver(value, theme) : value};`;
+  }
+
+  let first = true;
+
+  return css`
+    ${BREAKPOINT_ORDER.map(breakpoint => {
+      const v = value[breakpoint];
+      if (v === undefined) {
+        return undefined;
+      }
+
+      if (first) {
+        first = false;
+        return css`
+          @media (min-width: ${theme.breakpoints[breakpoint]}),
+            (max-width: ${theme.breakpoints[breakpoint]}) {
+            ${property}: ${resolver ? resolver(v, theme) : (v as string)};
+          }
+        `;
+      }
+
+      return css`
+        @media (min-width: ${theme.breakpoints[breakpoint]}) {
+          ${property}: ${resolver ? resolver(v, theme) : (v as string)};
+        }
+      `;
+    }).filter(Boolean)}
+  `;
 }
 
-export const Flex = styled('div', {
-  shouldForwardProp: prop =>
-    isPropValid(prop) &&
-    !['align', 'direction', 'flex', 'gap', 'inline', 'justify', 'wrap'].includes(prop),
-})<FlexProps>`
-  display: ${p => (p.inline ? 'inline-flex' : 'flex')};
-  flex-direction: ${p => p.direction};
-  justify-content: ${p => p.justify};
-  align-items: ${p => p.align};
+const omitFlexProps = new Set<keyof FlexProps>([
+  'as',
+  'direction',
+  'flex',
+  'gap',
+  'inline',
+  'align',
+  'justify',
```

Comment:
I’m a bit confused about why we have `align` instead of `align-items` and `justify` instead of `justify-content`. Is it because it’s influenced by `direction` ? Like, do we get `align-items` or `align-content` dependent on `direction`, because I don’t think we do ...

I’m not really a fan of introducing new APIs unless necessary or clearly beneficial. The problem is that props aren’t discoverable because we allow all html-props, too, so there’s a lot of them. And anything that’s different from standard flex naming needs to be learned anew...

Reply:
Resolved in DM, but it seems that this is already handled by the layout engine when direction is column, nice find @TkDodo 🙏🏼